// @unlink-xyz/node/dist/index.js
// src/init.ts
import crypto from "crypto";
import { createMemoryStorage, Unlink } from "@unlink-xyz/core";
var nodeRng = (n) => new Uint8Array(crypto.randomBytes(n));
async function initUnlink(config) {
  const { setup, sync, ...coreConfig } = config;
  void setup;
  void sync;
  const unlink = await Unlink.create({
    ...coreConfig,
    storage: config.storage ?? createMemoryStorage(),
    rng: config.rng ?? nodeRng,
    autoSync: config.autoSync ?? false,
    fetch: config.fetch ?? globalThis.fetch
  });
  if (config.setup !== false) {
    const hasSeed = await unlink.seed.exists();
    if (!hasSeed) {
      await unlink.seed.create();
    }
    const accounts = await unlink.accounts.list();
    if (accounts.length === 0) {
      await unlink.accounts.create();
    }
  }
  if (config.sync !== false) {
    await unlink.sync();
  }
  return unlink;
}
var initWallet = initUnlink;

// src/errors.ts
var TERMINAL_TX_STATES = [
  "succeeded",
  "reverted",
  "failed",
  "dead"
];
var FAILED_TX_STATES = [
  "reverted",
  "failed",
  "dead"
];
var DEFAULT_TIMEOUT_MS = 5 * 60 * 1e3;
var DEFAULT_POLL_INTERVAL_MS = 2e3;
var MAX_POLL_INTERVAL_MS = 3e4;
var BACKOFF_FACTOR = 1.5;
var TimeoutError = class extends Error {
  txId;
  timeout;
  constructor(txId, timeout) {
    super(`Transaction ${txId} did not confirm within ${timeout}ms`);
    this.name = "TimeoutError";
    this.txId = txId;
    this.timeout = timeout;
  }
};
var TransactionFailedError = class extends Error {
  txId;
  state;
  reason;
  constructor(txId, state, reason) {
    super(
      `Transaction ${txId} failed with state: ${state}${reason ? ` \u2014 ${reason}` : ""}`
    );
    this.name = "TransactionFailedError";
    this.txId = txId;
    this.state = state;
    this.reason = reason;
  }
};

// src/confirmation.ts
function toTxStatus(txId, raw) {
  return {
    txId,
    state: raw.state,
    txHash: raw.txHash ?? void 0,
    blockNumber: raw.receipt?.blockNumber ?? void 0,
    error: raw.error ?? void 0
  };
}
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
async function waitForConfirmation(wallet, txId, opts) {
  const timeout = opts?.timeout ?? DEFAULT_TIMEOUT_MS;
  let interval = opts?.pollInterval ?? DEFAULT_POLL_INTERVAL_MS;
  const deadline = Date.now() + timeout;
  while (Date.now() < deadline) {
    const raw = await wallet.getTxStatus(txId);
    const status = toTxStatus(txId, raw);
    if (status.state === "succeeded") {
      return status;
    }
    if (FAILED_TX_STATES.includes(status.state)) {
      throw new TransactionFailedError(txId, status.state, status.error);
    }
    const remaining = deadline - Date.now();
    if (remaining <= 0) break;
    await sleep(Math.min(interval, remaining));
    interval = Math.min(interval * BACKOFF_FACTOR, MAX_POLL_INTERVAL_MS);
  }
  throw new TimeoutError(txId, timeout);
}
async function* pollRelayStatus(wallet, txId, opts) {
  const timeout = opts?.timeout ?? DEFAULT_TIMEOUT_MS;
  let interval = opts?.pollInterval ?? DEFAULT_POLL_INTERVAL_MS;
  const deadline = Date.now() + timeout;
  while (Date.now() < deadline) {
    const raw = await wallet.getTxStatus(txId);
    const status = toTxStatus(txId, raw);
    yield status;
    if (TERMINAL_TX_STATES.includes(status.state)) {
      return;
    }
    const remaining = deadline - Date.now();
    if (remaining <= 0) return;
    await sleep(Math.min(interval, remaining));
    interval = Math.min(interval * BACKOFF_FACTOR, MAX_POLL_INTERVAL_MS);
  }
}

// src/events.ts
import { EventEmitter } from "events";
function createEmitter(unlink) {
  const emitter = new EventEmitter();
  const unsubscribe = unlink.on((event) => {
    emitter.emit(event.type, event);
    emitter.emit("*", event);
  });
  return { emitter, unsubscribe };
}
var createWalletEmitter = createEmitter;

// src/storage/sqlite.ts
import fs from "fs";
import path from "path";
import { validateKey } from "@unlink-xyz/core";
import Database from "better-sqlite3";
function createSqliteStorage(opts) {
  let db = null;
  let stmtGet;
  let stmtPut;
  let stmtDel;
  let stmtGetSchema;
  let stmtUpsertSchema;
  return {
    async open() {
      const dir = path.dirname(opts.path);
      fs.mkdirSync(dir, { recursive: true });
      db = new Database(opts.path);
      db.pragma("journal_mode = WAL");
      db.pragma("busy_timeout = 5000");
      db.exec(`
        CREATE TABLE IF NOT EXISTS kv (key TEXT PRIMARY KEY, value BLOB NOT NULL);
        CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value INTEGER NOT NULL);
      `);
      stmtGet = db.prepare("SELECT value FROM kv WHERE key = ?");
      stmtPut = db.prepare(
        "INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)"
      );
      stmtDel = db.prepare("DELETE FROM kv WHERE key = ?");
      stmtGetSchema = db.prepare(
        "SELECT value FROM meta WHERE key = 'schema_version'"
      );
      stmtUpsertSchema = db.prepare(
        "INSERT OR REPLACE INTO meta (key, value) VALUES ('schema_version', ?)"
      );
    },
    async get(key) {
      validateKey(key);
      const row = stmtGet.get(key);
      return row ? new Uint8Array(row.value) : null;
    },
    async put(key, value) {
      validateKey(key);
      stmtPut.run(key, Buffer.from(value));
    },
    async delete(key) {
      validateKey(key);
      stmtDel.run(key);
    },
    async batch(ops) {
      const runBatch = db.transaction(() => {
        for (const op of ops) {
          if (op.put) {
            const [k, v] = op.put;
            validateKey(k);
            stmtPut.run(k, Buffer.from(v));
          }
          if (op.del) {
            validateKey(op.del);
            stmtDel.run(op.del);
          }
        }
      });
      runBatch();
    },
    async iter(o = {}) {
      if (o.start && o.end && o.start > o.end) {
        throw new Error("iter start bound must not exceed end bound");
      }
      const conditions = [];
      const params = [];
      if (o.prefix) {
        const escaped = o.prefix.replace(/[%_\\]/g, "\\$&");
        conditions.push("key LIKE ? ESCAPE '\\'");
        params.push(escaped + "%");
      }
      if (o.start) {
        conditions.push("key >= ?");
        params.push(o.start);
      }
      if (o.end) {
        conditions.push("key <= ?");
        params.push(o.end);
      }
      let sql = "SELECT key, value FROM kv";
      if (conditions.length > 0) {
        sql += " WHERE " + conditions.join(" AND ");
      }
      sql += o.reverse ? " ORDER BY key DESC" : " ORDER BY key ASC";
      if (o.limit != null) {
        sql += " LIMIT ?";
        params.push(o.limit);
      }
      const rows = db.prepare(sql).all(...params);
      return rows.map((row) => ({
        key: row.key,
        value: new Uint8Array(row.value)
      }));
    },
    async count(prefix) {
      let sql = "SELECT COUNT(*) as cnt FROM kv";
      const params = [];
      if (prefix) {
        const escaped = prefix.replace(/[%_\\]/g, "\\$&");
        sql += " WHERE key LIKE ? ESCAPE '\\'";
        params.push(escaped + "%");
      }
      const row = db.prepare(sql).get(...params);
      return row.cnt;
    },
    async getSchemaVersion() {
      const row = stmtGetSchema.get();
      return row?.value ?? 0;
    },
    async setSchemaVersion(v) {
      stmtUpsertSchema.run(v);
    }
  };
}

// src/index.ts
import { Unlink as Unlink2, UnlinkWallet, createMemoryStorage as createMemoryStorage2 } from "@unlink-xyz/core";
export {
  DEFAULT_POLL_INTERVAL_MS,
  DEFAULT_TIMEOUT_MS,
  FAILED_TX_STATES,
  TERMINAL_TX_STATES,
  TimeoutError,
  TransactionFailedError,
  Unlink2 as Unlink,
  UnlinkWallet,
  createEmitter,
  createMemoryStorage2 as createMemoryStorage,
  createSqliteStorage,
  createWalletEmitter,
  initUnlink,
  initWallet,
  pollRelayStatus,
  waitForConfirmation
};
